python/ray/_private/workers/default_worker.py

```diff
@@ -270,15 +271,31 @@
         worker_launched_time_ms=worker_launched_time_ms,
     )
 
+    stdout_fileno = sys.stdout.fileno()
+    stderr_fileno = sys.stderr.fileno()
+    # We also manually set sys.stdout and sys.stderr because that seems to
+    # have an effect on the output buffering. Without doing this, stdout
+    # and stderr are heavily buffered resulting in seemingly lost logging
+    # statements. We never want to close the stdout file descriptor, dup2 will
+    # close it when necessary and we don't want python's GC to close it.
+    sys.stdout = ray._private.utils.open_log(
+        stdout_fileno, unbuffered=True, closefd=False
+    )
+    sys.stderr = ray._private.utils.open_log(
+        stderr_fileno, unbuffered=True, closefd=False
+    )
+
     worker = ray._private.worker.global_worker
 
     # Setup log file.
-    out_file, err_file = node.get_log_file_handles(
-        get_worker_log_file_name(args.worker_type)
+    out_filepath, err_filepath = node.get_log_file_names(
+        get_worker_log_file_name(args.worker_type),
+        unique=False,  # C++ core worker process already creates the file, should use a deterministic function to get the same file path.
+        create_out=True,
+        create_err=True,
     )
```

Comment:
We can let c++ core worker expose the stdout and stderr file path and python retrieves it via cython (see runtime context as examples) so we don't need to implement `get_worker_log_file_name` in python as well.

Reply:
> so we don't need to implement get_worker_log_file_name in python as well.

The function is used elsewhere so needed anyway.
```
[~/ray] (hjiang/readsome-redirect-core-worker-process) 
ubuntu@hjiang-devbox-pg$ git grep "get_worker_log_file_name("
python/ray/_private/ray_logging/__init__.py:def get_worker_log_file_name(worker_type, job_id=None):
python/ray/_private/workers/default_worker.py:        get_worker_log_file_name(args.worker_type),
python/ray/workflow/workflow_context.py:        get_worker_log_file_name("WORKER")
python/ray/workflow/workflow_context.py:        get_worker_log_file_name("WORKER", job_id)
```

Reply:
then can we make it a c++ function and expose to python to call via cython to avoid code duplication?

Reply:
I'm not sure it's a good idea, the simplified function here actually include a few number of functions in python.

https://github.com/ray-project/ray/blob/bff0bb4bf0bd8605588e4296145d0da8607aa5e3/python/ray/_private/ray_logging/__init__.py#L128-L146

https://github.com/ray-project/ray/blob/bff0bb4bf0bd8605588e4296145d0da8607aa5e3/python/ray/_private/node.py#L916-L942

https://github.com/ray-project/ray/blob/bff0bb4bf0bd8605588e4296145d0da8607aa5e3/python/ray/_private/node.py#L805-L839

https://github.com/ray-project/ray/blob/bff0bb4bf0bd8605588e4296145d0da8607aa5e3/python/ray/_private/utils.py#L94-L109

(the list goes on)

I don't think we need to we need to rewrite all these functions, but creating temp directory and support unique filename is a necessity.

Reply:
The `get_log_name` function is not an exact re-implementation for python version, we remove the repetition avoidance logic for core worker cases.

Reply:
OK a second thought, it's better to have a C++ implementation, so C++ related components' log names are all decided at C++ side (GCS/raylet/worker).

Three things needed here:
- A util function which re-implements python side logic, it would be more complicated in python since it involves cross-platform;
- Move the log filename decision logic completely into C++;
- Expose the util function to python.

Reply:
I leave a TODO item.

Reply:
But if we want to go this path, certain functions (i.e. `_make_inc_temp`) will duplicate in both languages anyway.

Reply:
For worker log, we don't need `_make_inc_temp` as worker id is part of the log and it's guaranteed to be unique.

Reply:
I need more clarification, isn't this thread talking about moving `get_log_fname` function into C++ but not both python + C++?

Since the function is used in python as well (i.e. log monitor's logging), we have to achieve the same effect in C++, which means `_make_inc_temp` needs to be written in C++ function as well.

Do I make myself clear?

Reply:
I mean at least `get_worker_log_file_name` can be moved to c++. We can discuss offline.

Reply:
I don't think so, you need to have another full implementation in python side.
A separate question, do you agree it's a non-trivial work and should be implemented in another PR?